File: plugins/ffi_c_plugin/lib/ffi_c_plugin.dart
import 'dart:ffi';
import 'dart:io';
import 'dart:ffi' as ffi;

import 'ffi_c_plugin_bindings_generated.dart';

const String _libName = 'ffi_c_plugin';

/// The dynamic library in which the symbols for [FfiCPluginBindings] can be found.
final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

void initC(
  int particlesAmount,
  double canvasWidth,
  double canvasHeight,
  double minMass,
  double maxMass,
) =>
    init_c(particlesAmount, canvasWidth, canvasHeight, minMass, maxMass);

ffi.Pointer<Particle> updateParticlesC() => update_particles_c();

File: plugins/ffi_c_plugin/src/ffi_c_plugin.cpp
#include "ffi_c_plugin.h"

#include <iostream>

NBody nbody = NBody(0, 0.0, 0.0, 0.0, 0.0);

void init_c(int32_t particles_amount,
          double canvas_width,
          double canvas_height,
          double min_mass,
          double max_mass)
{
    nbody = NBody(particles_amount,
                 canvas_width,
                 canvas_height,
                 min_mass,
                 max_mass);
    nbody.init();
}

Particle* update_particles_c() 
{
    nbody.update_particles();
    return nbody.particles.data();
}